massgov/mayflower
# [React/Patternlab]  DP-10983 Table of Contents

Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)


## Description

Based on the [accessibility audit](https://docs.google.com/spreadsheets/d/1cfKf8cLjZAFaIyVkAj1inCItXKMv0MynitFnaII1toY/edit#gid=1243348889), some improvements are implemented to the accordion components.
1. Removed `aria-label` from `button`, which provides a duplicate button label, and is not necessary.
2. Added pairing between `button` and `div.ma__accordion-content__body` with `aria-controls`.
3. Removed `<header class="ma__accordion-header">`, which is not semantically necessary.
4. Removed "+" symbol on the right from css `:after` since screenreaders read it as "plus" which is not relevant as the context. To keep the visual same, the symbol is inserted in `span` with `aria-hidden=true`, which is not necessary to be read by screenreaders since `aria-expanded` tells the status.
5. Took out `heading` nested in the `button`, which was invalid makeup (headings are not allowed in button), placed the `heading` at where <header class="ma__accordion-header">` used to be. In another words, use the heading as a wrapper for the button.
6. Add a condition to render the heading as necessary.  Currently, an accordion item always gets heading, but there are cases the heading is not necessary. The button can be rendered with or without heading.
7. Apply the same changes to Patternlab.

## Related Issue / Ticket

- [DP-10983](https://jira.mass.gov/browse/DP-10983)

## Steps to Test


### React
1. Go to Table of Contents with ColoredHeading under organisms in Storybook. In the source, check:
    - Accordion item container `div.ma__accordion-item` doesn't have `<header>`.
    - The accordion button is wrapped with the heading with its proper heading level (3, note the colored heading `.ma__colored-heading` is `h2`.) 
     -  The heading has a class `ma__accordion-header`.
     -  The button has NO `aria-label`.
     -  The button has `aria-controls` with its value as the `id` value of its corresponding `div.ma__accordion-content__body`.

2. Find "+" symbol at the right of the **closed** accordion button.

3. Find "x" symbol at the right of the **open** accordion button.

### Patternlab
There are two templates to use this accordion pattern:
- Collapsible Header
- Collapsible Content as Contacts

1. Go to http://localhost:3000/patterns/03-organisms/index.html?_page=1.  Here you can find both samples.

3. Scroll down to "Collapsible Content as Steps.  Please be aware this heading (.ma__comp-heading) is level 2.  Check the source of `.ma__collapsible-content__items`.
     - In the `div.ma__collapsible-content__items`, the next wrapper is `div.ma__collapsible-content__item` followed by `h3.ma__collapsible-header`.  No `header.ma__collapsible-header` in the item.
     - In this context, find the heading level is semantically correct.  The heading level is adjustable based on its parent heading level.
     - Find the button label is in `span.ma__collapsible-header__title`, which used to be `h2`.

2. Scroll down to "Contacts".  Please be aware this heading (.ma__comp-heading) is level 2.  Check the source of `section.ma__contact-us`.
     - The `section.ma__contact-us` is followed by `h3.ma__column-heading`.
     - In this context, find the heading level is semantically correct.  The heading level is adjustable based on its parent heading level.

### Test with Voiceover (Mac only)
1. Turn on Voiceover with **command + F5**.
2. Click the area to get focus.
3. Use **Control + option + Shift + RIGHT arrow key** to move forward, **Control + option + Shift + LEFT arrow key** to move backward.
4. When you are on the accordion button, use **space** to click to open the accordion. Hit **space** again to close.
5. Check for:
     - The accordion button label is read.
     - The status of the accordion (collapsed/expanded) is announced.
     - As clicking the "+/x" symbol, the animation works.
     - Reading order makes sense. 

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.


## Additional Notes:

Anything else to add?

#### Impacted Areas in Application


* 

#### @TODO


*

#### Today I learned...
